rust/sdk-processor/src/processors/events_processor.rs

```diff
@@ -0,0 +1,120 @@
+};
+use std::time::Duration;
+use tracing::{debug, info};
+
+pub struct EventsProcessor {
+    pub config: IndexerProcessorConfig,
+    pub postgres_config: PostgresConfig,
+    pub db_pool: ArcDbPool,
+}
+
+impl EventsProcessor {
+    pub async fn new(config: IndexerProcessorConfig) -> Result<Self> {
+        let db_config = config.db_config.clone();
+        match db_config {
+            DbConfig::PostgresConfig(postgres_db_config) => {
+                let conn_pool = new_db_pool(
+                    &postgres_db_config.connection_string,
+                    Some(postgres_db_config.db_pool_size),
+                )
+                .await
+                .map_err(|e| {
+                    anyhow::anyhow!(
+                        "Failed to create connection pool for PostgresConfig: {:?}",
+                        e
+                    )
+                })?;
+
+                Ok(Self {
+                    config,
+                    postgres_config: postgres_db_config.clone(),
+                    db_pool: conn_pool,
+                })
+            },
+        }
+    }
+
+    pub async fn run_processor(self) -> Result<()> {
+        // (Optional) Run migrations
+        run_migrations(
+            self.postgres_config.connection_string.clone(),
+            self.db_pool.clone(),
+        )
+        .await;
+
+        // (Optional) Merge the starting version from config and the latest processed version from the DB
+        let starting_version = get_starting_version(&self.config, self.db_pool.clone()).await?;
+
+        // Define processor steps
+        let (_input_sender, input_receiver) = instrumented_bounded_channel("input", 1);
+
+        let transaction_stream = TransactionStreamStep::new(TransactionStreamConfig {
+            starting_version: Some(starting_version),
+            ..self.config.transaction_stream_config
+        })
+        .await?;
+        let transaction_stream_with_input = RunnableStepWithInputReceiver::new(
+            input_receiver,
+            transaction_stream.into_runnable_step(),
+        );
+        let events_extractor = EventsExtractor {};
+        let events_storer = EventsStorer::new(self.db_pool.clone(), self.postgres_config.clone());
+        let timed_buffer = TimedBuffer::new(Duration::from_secs(1));
+        let version_tracker = LatestVersionProcessedTracker::new(
+            self.db_pool.clone(),
+            starting_version,
+            self.config.processor_config.name().to_string(),
+        );
+
+        // Connect processor steps together
+        let (_, buffer_receiver) = ProcessorBuilder::new_with_runnable_input_receiver_first_step(
+            transaction_stream_with_input,
+        )
+        .connect_to(events_extractor.into_runnable_step(), 10)
+        .connect_to(timed_buffer.into_runnable_step(), 10)
+        .connect_to(events_storer.into_runnable_step(), 10)
+        .connect_to(version_tracker.into_runnable_step(), 10)
```

Comment:
Is there any step that process txns out of order? If not can we have the version tracker to enforce this?

Reply:
We don't have this now, but max was saying there's a way to create worker tasks that each pull from the same channel and run the same step.